mitodl/open-discussions
# Blocked IP model and middleware

#### Pre-Flight checklist


#### What are the relevant tickets?
Closes #3070 

#### What's this PR do?
- Adds a `BlockedIPRange` model
- Adds middleware to block POST/PUT/PATCH requests if coming from a user IP in a blocked range (if the IP is routable and the user is not a superuser).
- Adds a `blocked_ip_exempt` decorator so elasticsearch requests are not affected.

#### How should this be manually tested?
- As a superuser, add a new `BlockedIPRange` object that includes your current IP (likely 10.something-something). You can also try entering some invalid values (null, non-ip values, end < start) and verify that the validation works.
- You should still be able to do everything (including create posts/comments).
- In an incognito browser, try logging in as a non-superuser and registering as a new user.  You should not be able to.
- Try searching for courses, podcasts, etc.  You should still be able to do that as an anonymous user.

#### Screenshots (if appropriate)
<img width="1069" alt="Screen Shot 2020-07-28 at 4 36 29 PM" src="https://user-images.githubusercontent.com/187676/88718981-917f7800-d0f0-11ea-82b9-2ac2078e3631.png">


<img width="1082" alt="Screen Shot 2020-07-28 at 4 36 13 PM" src="https://user-images.githubusercontent.com/187676/88718995-9512ff00-d0f0-11ea-8e69-8ccd35bb9861.png">
